lib/semaphore.ml

```diff
@@ -0,0 +1,68 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+type semaphore = {
+  mutable n : int;
+  m : Mutex.t;
+  c : Condition.t;
+}
+
+let create n =
+  if n <= 0 then
+    invalid_arg (Printf.sprintf 
+                   "Semaphore value must be positive, got %d" n);
+  let m = Mutex.create ()
+  and c = Condition.create () in
+  { n; m; c; }
+
+exception Inconsistent_state of string
+let inconsistent_state string = raise (Inconsistent_state string)
```

Comment:
It's convenient to allow printf-style formats: `let f fmt = Printf.kprintf (fun msg -> raise (Inconsistent_state msg) fmt`